Run remove_skip and goto_program.update() in passes that may introduce skips

tautschnig opened the pull request:
The only passes left out are set_properties and remove_unreachable as the caller may still find value in those instructions replaced by a skip. (This need not be true and it may as well be safe to run remove_skip for those as well.)

tautschnig committed:
Use make_skip to turn an instruction into a SKIP

tautschnig committed:
Fix wrong function description

tautschnig committed:
Remove redundant calls to compute_{location,loop}_numbers

tautschnig committed:
Run remove_skip in passes that may introduce skips

The only passes left out are set_properties and remove_unreachable as the caller
may still find value in those instructions replaced by a skip. (This need not be
true and it may as well be safe to run remove_skip for those as well.)

Fixes: #2067

smowton commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -235,6 +232,12 @@ int goto_instrument_parse_optionst::doit()
+        // after having generated the log above
+        remove_skip(goto_model);
+        goto_model.goto_functions.update();
+        goto_model.goto_functions.compute_loop_numbers();
```
Redundant: update runs compute_loop_numbers

smowton commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -553,6 +550,9 @@ void remove_virtual_function(
   rvf.remove_virtual_function(
     goto_program, instruction, dispatch_table, fallback_action);
+
+  remove_skip(goto_program);
```
I don't think the singular version should do this -- it's supposed to be lowering a single call into a dispatch table, which should not imply changing unrelated parts of the goto program.

tautschnig replied:
Point taken, it might be necessary to come up with a version of `remove_skip` that just takes a range of instructions.

smowton replied:
I think it would suffice to skip doing this in this case -- the convention is that a *pass* should leave the location numbers and GOTO program in a "tidy" state, and this particular flavour of RVF is not a pass, it's a utility for other passes.

tautschnig replied:
Thanks, I'll do a bit more code reading and will then likely just put a comment in there.

smowton reviewed (commented):
Looks ok except one case which I think is the wrong move.

Will make a TG bump.

smowton commented:
https://github.com/diffblue/test-gen/pull/1729

smowton commented:
@tautschnig one other thing -- should the same PR also ensure any pass that *adds* instructions calls `update()`?

tautschnig commented:
I will review the use/absence of `update()` and will confirm once done.

tautschnig committed:
remove_skip implementation restricted to a range of instructions

tautschnig committed:
remove_skip includes goto_{program,functions}.update(), avoid redundant calls

tautschnig committed:
Avoid redundant calls to remove_skip

tautschnig committed:
Remove unused includes of remove_skip.h

tautschnig committed:
Make remove_skip call goto_program.update()

All invocations of remove_skip do this immediately after calling remove_skip,
which just complicates the protocol of using remove_skip.

tautschnig committed:
Explicitly invoke goto_program.update() where remove_skip is not used

tautschnig commented:
I've made a number of updates; once CI passes here this will definitively need confirmation that the test-gen bump (still) passes.

smowton commented:
@tautschnig updated it just now

smowton commented:
It failed, but looks like this doesn't actually compile by itself at present, so I'll leave it until this is updated.

tautschnig committed:
Enable remove_skip to soundly remove skips with labels

Also add documentation to what is_skip actually does.

tautschnig commented:
@smowton Finally fixed. Would you mind reviving the TG bump?

smowton commented:
@tautschnig updated

smowton commented:
Bump failed due to concurrent moving of remove_java_nondet.h. Rebased and retrying just now.

smowton commented:
@tautschnig bump is passing

tautschnig commented:
Thanks a lot for confirming! Now I'd just need reviews/approvals...

thk123 commented on src/goto-instrument/model_argc_argv.cpp:
```diff
@@ -181,7 +181,6 @@ bool model_argc_argv(
   // update counters etc.
   remove_skip(start);
-  start.compute_loop_numbers();
   goto_model.goto_functions.update();
```
Isn't this now redundant too?

tautschnig replied:
Done.

thk123 commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -283,13 +289,14 @@ void remove_virtual_functionst::remove_virtual_function(
   // finally, kill original invocation
   target->make_skip();
+
+  remove_skip(goto_program, target, next_target);
```
Here we are removing skips from the virtual function block? Consider commenting that's what the `[target, next_target]` range corresponds to

tautschnig replied:
Done.

thk123 commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -472,23 +479,29 @@ bool remove_virtual_functionst::remove_virtual_functions(
+  for(goto_programt::instructionst::iterator
+      target = goto_program.instructions.begin();
+      target != goto_program.instructions.end();
+      ) // no ++target
```
Nit: this comment is confusing as it looks like `++target`!! Consider `// Don't increment target (why ??)`

tautschnig replied:
Done.

thk123 commented on src/goto-programs/remove_skip.cpp:
```diff
@@ -12,7 +12,10 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "goto_model.h"
 
-bool is_skip(const goto_programt &body, goto_programt::const_targett it)
+bool is_skip(
```
Please document the new param at least

tautschnig replied:
Thank you for making me do this. It helped me see that I hadn't actually implemented this new feature... oO - fixed now.

thk123 reviewed (commented):
It looks essentially fine - though it seemed like the passes you were adding the remove_skip call you didn't have a corresponding remove? 

I also think @smowton bump isn't pointing to the right branch atm

smowton commented:
Needs rebase and then re-test, but as of last check the test-gen bump was ok

tautschnig commented:
@peterschrammel Comments addressed and rebased.

tautschnig committed:
Relax test: there need not be any code at the end of the block to cover

tautschnig committed:
Make functions non-empty so that the test remains stable under skip cleanup

tautschnig commented:
Updated against current HEAD and updated tests as needed.

tautschnig commented:
Now also fixed a goto-diff test. This absolutely needs a TG bump, there may be impact on tests using empty functions.

peterschrammel commented:
TG bump updated.

peterschrammel commented on regression/goto-diff/java-properties/test.desc:
```diff
@@ -5,6 +5,6 @@ old.jar --json-ui --show-properties --cover location
 EXIT=0
 SIGNAL=0
-  "deletedFunctions": \[\n      \{\n        "name": "java::Test.obsolete:\(\)V",\n        "properties": \[\n          \{\n            "class": "coverage",\n            "coveredLines": "18",\n            "description": "block 1",\n            "expression": "false",\n            "name": "java::Test.obsolete:\(\)V.coverage.1",\n            "sourceLocation": \{\n              "bytecodeIndex": "0",\n  ...
+  "deletedFunctions": \[\n      \{\n        "name": "java::Test.obsolete:\(\)V",\n        "properties": \[ \],\n        "sourceLocation": \{\n          "file": "Test.java",\n          "function": "java::Test.obsolete:\(\)V",\n          "line": "18"\n        }\n      }\n    \],\n    "modifiedFunctions": \[\n      \{\n        "name": "java::Test.<init>:\(\)V",\n        "properties": \[\n          \ ...
```
This is certainly not the desired behaviour. Even a goto function that consists only of END_FUNCTION should get a coverage goal added.

tautschnig replied:
Indeed removing labelled `SKIP`s when those labels were transferred to `END_FUNCTION` was wrong. This is now fixed (and those changes have thus been removed).

martin-cs commented on src/goto-programs/remove_skip.cpp:
```diff
@@ -12,12 +12,22 @@ Author: Daniel Kroening, kroening@kroening.com
+/// \param ignore_labels  If the caller takes care of moving labels, then even
+/// skip statements carrying labels can be treated as skips (even though they
+/// may carry key information such as error labels).
+/// \return True, iff it is equivalent to a skip.
```
I'm having a stuck record moment but ... what guarantees do our tools provide on the preservation of locations, instruction IDs, etc. through various transforms?  For example, if I am an entirely hypothetical commercial user of CPROVER, how do I map results back through transformation steps?

This is not an issue specifically or directly to do with this PR (although I think this PR may make it more involved) but I think it is something that does need to be considered at some point / in some way.

martin-cs reviewed (approved):
Nothing blocking; just philosophy.

smowton commented:
Looks like if you wanted to perturb current behaviour less you could just restrict your remove-labelled-skip work to never remove a skip that immediately precedes the end of the function, as this tends to throw away source location information and therefore a coverage target. Alternatively, have the Java frontend route return statements via a `LOCATION` instruction that ensures something is retained at the end of each function.

peterschrammel commented:
Rebased the TG bump to check.

peterschrammel reviewed (approved):
TG bump passed.
